configcatclient/rolloutevaluator.py

```diff
@@ -48,174 +169,553 @@ def evaluate(self, key, user, default_value, default_variation_id, settings):  #
+            for comparison in comparison_value:
+                underscore_index = comparison.index('_')
+                length = int(comparison[:underscore_index])
 
-    def _format_match_rule(self, comparison_attribute, user_value, comparator, comparison_value, value):
-        return 'Evaluating rule: [%s:%s] [%s] [%s] => match, returning: %s' \
-               % (comparison_attribute, user_value, self.COMPARATOR_TEXTS[comparator], comparison_value, value)
+                if len(user_value) >= length:
+                    comparison_string = comparison[underscore_index + 1:]
+                    if (
+                            (comparator == Comparator.STARTS_WITH_ANY_OF_HASHED
+                             and sha256(user_value[:length], salt, context_salt) == comparison_string)
+                            or
+                            (comparator == Comparator.NOT_STARTS_WITH_ANY_OF_HASHED
+                             and sha256(user_value[:length], salt, context_salt) != comparison_string)
+                            or
+                            (comparator == Comparator.ENDS_WITH_ANY_OF_HASHED
+                             and sha256(user_value[-length:], salt, context_salt) == comparison_string)
+                            or
+                            (comparator == Comparator.NOT_ENDS_WITH_ANY_OF_HASHED
+                             and sha256(user_value[-length:], salt, context_salt) != comparison_string)
+                    ):
+                        return True, error
+        # ARRAY CONTAINS ANY OF, ARRAY NOT CONTAINS ANY OF (hashed)
+        elif Comparator.ARRAY_CONTAINS_ANY_OF_HASHED <= comparator <= Comparator.ARRAY_NOT_CONTAINS_ANY_OF_HASHED:
+            try:
+                user_value_list = json.loads(user_value)
+                if not isinstance(user_value_list, list):
+                    raise ValueError()
+            except ValueError:
+                validation_error = "'%s' is not a valid JSON string array" % str(user_value)
+                error = self._handle_invalid_user_attribute(comparison_attribute, comparator, comparison_value, key,
+                                                            validation_error)
+                return False, error
 
-    def _format_no_match_rule(self, comparison_attribute, user_value, comparator, comparison_value):
-        return 'Evaluating rule: [%s:%s] [%s] [%s] => no match' \
-               % (comparison_attribute, user_value, self.COMPARATOR_TEXTS[comparator], comparison_value)
+            for comparison in comparison_value:
+                if comparator == Comparator.ARRAY_CONTAINS_ANY_OF_HASHED:
+                    if comparison in [sha256(x.strip(), salt, context_salt) for x in user_value_list]:
+                        return True, error
+                else:
+                    if comparison not in [sha256(x.strip(), salt, context_salt) for x in user_value_list]:
```

Comment:
I'm not 100% sure, but I think this will not work as expected if there are multiple values on the dashboard. I asked for some tests around this in the SDK channel.
Dashboard: if user attribute array not contains any of [a,b,c]
user object value: [e,f,b]
I think this should evaluate to false, but this code runs through the [a,b,c] array and if the first value is not in the user provided array, it returns true.
I'm not 100% sure, let's see after the new test cases.